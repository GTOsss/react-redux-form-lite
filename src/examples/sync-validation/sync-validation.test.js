import React from 'react';
import {mount} from 'enzyme';
import {Provider} from 'react-redux';
import {createStore, applyMiddleware, combineReducers} from 'redux';
import SyncValidation from './sync-validation';
import ReduxThunkTester from 'redux-thunk-tester';
import {reducer} from '../../index';

const renderComponent = (onSubmit) => {
  const reduxThunkTester = new ReduxThunkTester();

  const store = createStore(
    combineReducers({reduxForm: reducer}),
    applyMiddleware(reduxThunkTester.createReduxThunkHistoryMiddleware()),
  );

  const component = mount(<Provider store={store}><SyncValidation onSubmit={onSubmit}/></Provider>);

  return {reduxThunkTester, store, component};
};

describe('<SyncValidation />', () => {
  test('render sync validation form', () => {
    const {component} = renderComponent(() => {});
    expect(component).toMatchSnapshot();
  });

  test('validate', () => {
    const {component, store} = renderComponent(() => {});

    component.find('input').at(0).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');

    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('validate: onSubmit after failed submit', () => {
    const onSubmit = jest.fn();
    const {component, store} = renderComponent(onSubmit);

    component.find('input').at(0).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    expect(onSubmit.mock.calls.length).toBe(1);
    expect(onSubmit.mock.calls[0][0]).toEqual({firstName: 'test', lastName: undefined});
    expect(onSubmit.mock.calls[0][1].form.hasErrors).toBeTruthy();
    expect(onSubmit.mock.calls[0][1].form.hasWarnings).toBeFalsy();
    expect(onSubmit.mock.calls[0][1].form.errorsMap).toEqual({lastName: 'Field required.'});
    expect(onSubmit.mock.calls[0][1].form.warningsMap).toEqual({});
    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });
  
  test('validate: onSubmit after failed and success', () => {
    const onSubmit = jest.fn();
    const {component, store} = renderComponent(onSubmit);

    component.find('input').at(0).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    expect(onSubmit.mock.calls.length).toBe(2);
    expect(onSubmit.mock.calls[1][0]).toEqual({firstName: 'test', lastName: 'test'});
    expect(onSubmit.mock.calls[1][1].form.hasErrors).toBeFalsy();
    expect(onSubmit.mock.calls[1][1].form.hasWarnings).toBeFalsy();
    expect(onSubmit.mock.calls[1][1].form.errorsMap).toEqual({});
    expect(onSubmit.mock.calls[1][1].form.warningsMap).toEqual({});
    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('warn', () => {
    const {component, store} = renderComponent(() => {});

    component.find('input').at(0).simulate('change', {target: {value: 't'}});
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');

    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('warn: onSubmit after failed submit', () => {
    const onSubmit = jest.fn();
    const {component, store} = renderComponent(onSubmit);

    component.find('input').at(0).simulate('change', {target: {value: 't'}});
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    expect(onSubmit.mock.calls.length).toBe(1);
    expect(onSubmit.mock.calls[0][0]).toEqual({firstName: 't', lastName: 'test'});
    expect(onSubmit.mock.calls[0][1].form.hasErrors).toBeFalsy();
    expect(onSubmit.mock.calls[0][1].form.hasWarnings).toBeTruthy();
    expect(onSubmit.mock.calls[0][1].form.warningsMap).toEqual({
      firstName: 'The minimum length of the value must be 2.',
    });
    expect(onSubmit.mock.calls[0][1].form.errorsMap).toEqual({});
    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('warn: onSubmit after failed and success', () => {
    const onSubmit = jest.fn();
    const {component, store} = renderComponent(onSubmit);

    component.find('input').at(0).simulate('change', {target: {value: 't'}});
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    component.find('input').at(0).simulate('change', {target: {value: 'test'}});
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    expect(onSubmit.mock.calls.length).toBe(2);
    expect(onSubmit.mock.calls[1][0]).toEqual({firstName: 'test', lastName: 'test'});
    expect(onSubmit.mock.calls[1][1].form.hasErrors).toBeFalsy();
    expect(onSubmit.mock.calls[1][1].form.hasWarnings).toBeFalsy();
    expect(onSubmit.mock.calls[1][1].form.errorsMap).toEqual({});
    expect(onSubmit.mock.calls[1][1].form.warningsMap).toEqual({});
    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });
});
