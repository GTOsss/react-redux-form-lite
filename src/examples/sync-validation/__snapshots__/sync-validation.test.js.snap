// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SyncValidation /> render sync validation form 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectFunction
    onSubmit={[Function]}
  >
    <ReduxForm
      actions={
        Object {
          "arrayPush": [Function],
          "blur": [Function],
          "change": [Function],
          "changeSubmitted": [Function],
          "focus": [Function],
          "registerField": [Function],
          "registerForm": [Function],
          "removeField": [Function],
          "updateErrorAndWarningMessages": [Function],
          "updateFormState": [Function],
          "updateValidateMessage": [Function],
          "updateValidateMessages": [Function],
          "updateWarningMessage": [Function],
          "updateWarningMessages": [Function],
        }
      }
      form={
        Object {
          "form": Object {
            "activeField": "",
            "blurred": false,
            "changed": false,
            "errorsMap": Object {},
            "focused": false,
            "hasErrors": false,
            "hasWarnings": false,
            "submitted": false,
            "warningsMap": Object {},
          },
          "meta": Object {
            "firstName": Object {
              "active": false,
              "blurred": false,
              "changed": false,
              "error": "",
              "focused": false,
              "warning": "",
            },
            "lastName": Object {
              "active": false,
              "blurred": false,
              "changed": false,
              "error": "",
              "focused": false,
              "warning": "",
            },
          },
          "values": Object {
            "firstName": undefined,
            "lastName": undefined,
          },
        }
      }
      ownProps={
        Object {
          "onSubmit": [Function],
        }
      }
    >
      <InnerSyncValidation
        form={
          Object {
            "form": Object {
              "activeField": "",
              "blurred": false,
              "changed": false,
              "errorsMap": Object {},
              "focused": false,
              "hasErrors": false,
              "hasWarnings": false,
              "submitted": false,
              "warningsMap": Object {},
            },
            "meta": Object {
              "firstName": Object {
                "active": false,
                "blurred": false,
                "changed": false,
                "error": "",
                "focused": false,
                "warning": "",
              },
              "lastName": Object {
                "active": false,
                "blurred": false,
                "changed": false,
                "error": "",
                "focused": false,
                "warning": "",
              },
            },
            "values": Object {
              "firstName": undefined,
              "lastName": undefined,
            },
          }
        }
        formActions={
          Object {
            "arrayPush": [Function],
            "blur": [Function],
            "change": [Function],
            "changeSubmitted": [Function],
            "focus": [Function],
            "registerField": [Function],
            "registerForm": [Function],
            "removeField": [Function],
            "updateErrorAndWarningMessages": [Function],
            "updateFormState": [Function],
            "updateValidateMessage": [Function],
            "updateValidateMessages": [Function],
            "updateWarningMessage": [Function],
            "updateWarningMessages": [Function],
          }
        }
        formParams={
          Object {
            "destroyOnUnmount": true,
            "form": "example",
          }
        }
        handleSubmit={[Function]}
        onSubmit={[Function]}
      >
        <form
          onSubmit={[Function]}
        >
          <div>
            <label>
              First Name
            </label>
            <div>
              <FieldWithContext
                component="input"
                name="firstName"
                placeholder="First Name"
                type="text"
              >
                <ConnectFunction
                  component="input"
                  formContext={
                    Object {
                      "destroyOnUnmount": true,
                      "form": "example",
                      "updateValidateAndWarningMap": [Function],
                      "validate": [Function],
                      "warn": [Function],
                    }
                  }
                  name="firstName"
                  placeholder="First Name"
                  type="text"
                >
                  <Field
                    actions={
                      Object {
                        "arrayPush": [Function],
                        "blur": [Function],
                        "change": [Function],
                        "changeSubmitted": [Function],
                        "focus": [Function],
                        "registerField": [Function],
                        "registerForm": [Function],
                        "removeField": [Function],
                        "updateErrorAndWarningMessages": [Function],
                        "updateFormState": [Function],
                        "updateValidateMessage": [Function],
                        "updateValidateMessages": [Function],
                        "updateWarningMessage": [Function],
                        "updateWarningMessages": [Function],
                      }
                    }
                    component="input"
                    fieldArrayContext={Object {}}
                    formContext={
                      Object {
                        "destroyOnUnmount": true,
                        "form": "example",
                        "updateValidateAndWarningMap": [Function],
                        "validate": [Function],
                        "warn": [Function],
                      }
                    }
                    name="firstName"
                    placeholder="First Name"
                    type="text"
                  >
                    <input
                      name="firstName"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      placeholder="First Name"
                      type="text"
                    />
                  </Field>
                </ConnectFunction>
              </FieldWithContext>
            </div>
          </div>
          <div>
            <label>
              Last Name
            </label>
            <div>
              <FieldWithContext
                component="input"
                name="lastName"
                placeholder="Last Name"
                type="text"
              >
                <ConnectFunction
                  component="input"
                  formContext={
                    Object {
                      "destroyOnUnmount": true,
                      "form": "example",
                      "updateValidateAndWarningMap": [Function],
                      "validate": [Function],
                      "warn": [Function],
                    }
                  }
                  name="lastName"
                  placeholder="Last Name"
                  type="text"
                >
                  <Field
                    actions={
                      Object {
                        "arrayPush": [Function],
                        "blur": [Function],
                        "change": [Function],
                        "changeSubmitted": [Function],
                        "focus": [Function],
                        "registerField": [Function],
                        "registerForm": [Function],
                        "removeField": [Function],
                        "updateErrorAndWarningMessages": [Function],
                        "updateFormState": [Function],
                        "updateValidateMessage": [Function],
                        "updateValidateMessages": [Function],
                        "updateWarningMessage": [Function],
                        "updateWarningMessages": [Function],
                      }
                    }
                    component="input"
                    fieldArrayContext={Object {}}
                    formContext={
                      Object {
                        "destroyOnUnmount": true,
                        "form": "example",
                        "updateValidateAndWarningMap": [Function],
                        "validate": [Function],
                        "warn": [Function],
                      }
                    }
                    name="lastName"
                    placeholder="Last Name"
                    type="text"
                  >
                    <input
                      name="lastName"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      placeholder="Last Name"
                      type="text"
                    />
                  </Field>
                </ConnectFunction>
              </FieldWithContext>
            </div>
          </div>
          <div>
            <button
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </InnerSyncValidation>
    </ReduxForm>
  </ConnectFunction>
</Provider>
`;

exports[`<SyncValidation /> validate 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {
      "lastName": "Field required.",
    },
    "focused": false,
    "hasErrors": true,
    "hasWarnings": false,
    "submitted": true,
    "warningsMap": Object {},
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": false,
      "error": "Field required.",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "test",
    "lastName": undefined,
  },
}
`;

exports[`<SyncValidation /> validate: onSubmit after failed and success 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {},
    "focused": false,
    "hasErrors": false,
    "hasWarnings": false,
    "submitted": true,
    "warningsMap": Object {},
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "test",
    "lastName": "test",
  },
}
`;

exports[`<SyncValidation /> validate: onSubmit after failed submit 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {
      "lastName": "Field required.",
    },
    "focused": false,
    "hasErrors": true,
    "hasWarnings": false,
    "submitted": true,
    "warningsMap": Object {},
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": false,
      "error": "Field required.",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "test",
    "lastName": undefined,
  },
}
`;

exports[`<SyncValidation /> warn 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {},
    "focused": false,
    "hasErrors": false,
    "hasWarnings": true,
    "submitted": true,
    "warningsMap": Object {
      "firstName": "The minimum length of the value must be 2.",
    },
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "The minimum length of the value must be 2.",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "t",
    "lastName": "test",
  },
}
`;

exports[`<SyncValidation /> warn: onSubmit after failed and success 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {},
    "focused": false,
    "hasErrors": false,
    "hasWarnings": false,
    "submitted": true,
    "warningsMap": Object {},
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "test",
    "lastName": "test",
  },
}
`;

exports[`<SyncValidation /> warn: onSubmit after failed submit 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {},
    "focused": false,
    "hasErrors": false,
    "hasWarnings": true,
    "submitted": true,
    "warningsMap": Object {
      "firstName": "The minimum length of the value must be 2.",
    },
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "The minimum length of the value must be 2.",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "t",
    "lastName": "test",
  },
}
`;
